include/coap2/resource.h

```diff
@@ -38,12 +38,10 @@
  * Definition of message handler function
  */
 typedef void (*coap_method_handler_t)
-  (coap_context_t  *,
-   coap_resource_t *,
+  (coap_resource_t *,
    coap_session_t *,
-   coap_pdu_t *,
-   coap_binary_t * /* token */,
-   coap_string_t * /* query string */,
+   const coap_pdu_t * /* request */,
+   const coap_string_t * /* query string */,
```

Comment:
Why keep the query?

Reply:
The handler may be called from an observe trigger, in which case there is no request PDU, but the Query may be needed.

I have been thinking about keeping the PDU for any observe subscription - from which the Query can be abstracted.  The benefit of this is that a FETCH observe with its data will also be kept in the PDU and so can be properly supported in the handler.

I cant do much more on this today, but converting a subscription to also hold the PDU is straightforward and would not require that much extra ram - but will need to update the lwip  / contiki PDU pool counts.

Reply:
The RAM requirements would depend on the request's size (FETCH + large body). But this could be handled by setting a configurable upper limit. If we did this, how would the handler distinguish between an original request and the trigger?

Reply:
With this PR (so the token could be removed), the observe trigger logic that sets up the response PDU adds in the Observe Option (as does handle_request() on a new subscription) so there is no need for the app response handler or the coap_add_data_large_response() / coap_add_data_blocked_response() logic do not need to add in the Observe option.  Again taking away for the knowledge of what is going on from the app layer.

So, I do not think the handler needs to know whether this is the initial or observe triggered request.

Reply:
Good. In this case I suggest that I continue reviewing this PR, and we may decide whether to elicit the query string from the callback handler after merge.

Reply:
Sure - will wait until you have finished reviewing this PR.

The saving of the PDU for Observe trigger will follow the same logic as saving the PDU for async delayed responses.